app/views/vm_common/_snap.html.haml

```diff
@@ -1,40 +1,7 @@
-= form_tag({:action => 'snap_vm',
-            :id     => @record},
-           :id     => "vm_snap_form",
-           :method => :post) do
-  = render :partial => "layouts/flash_msg"
-  #form_div
-    %h3= _('Snapshot Information')
-    .form-horizontal
-      %div{:class => "form-group #{@record.vendor == "redhat" ? "hidden" : ""}"}
-        %label.control-label.col-md-2
-          = _('Name')
-        .col-md-8
-          = text_field_tag("name", @name, :maxlength => ViewHelper::MAX_NAME_LEN, :class => "form-control", "data-miq_focus" => true)
-      .form-group
-        %label.control-label.col-md-2
-          = _('Description')
-        .col-md-8
-          = text_area_tag("description", @description, :size => "50x4")
-      %div{:class => "form-group #{@show_snapshot_memory_checkbox ? "" : "hidden"}"}
-        %label.control-label.col-md-2
-          = _('Snapshot VM memory')
-        .col-md-8
-          = check_box_tag("snap_memory", "1", false)
-  - unless @edit
-    %hr
-    %table{:width => "100%"}
-      %tr
-        %td{:align => "right"}
-          = button_tag(_("Create"),
-            :name  => "create",
-            :class => "button",
-            :alt   => (t = _("Create snapshot")),
-            :title => t,
-            :type  => "submit")
-          = button_tag((t = _("Cancel")),
-            :name  => "cancel",
-            :class => "button",
-            :alt   => t,
-            :title => t,
-            :type  => "submit")
+= render :partial => "layouts/flash_msg"
+#snap-form{"data-create-url" => url_for_only_path(:action => "snap_vm", :id => @record.id, :button => @edit ? "create" : nil),
+           "data-cancel-url" => url_for_only_path(:action => "snap_vm", :id => @record.id, :button => @edit ? "cancel" : nil),
+           "data-name-required"  => !!@record.try(:snapshot_name_optional?)}
+
+:javascript
+  ManageIQ.component.componentFactory('VmSnapshotForm', 'snap-form');
```

Comment:
why not simply pass json instead of an html? am i missing something or is this what https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/helpers/reactjs_helper.rb suppose to be doing.....?

Reply:
I think its better to do it in ruby vs doing in html template (haml) if its going to be consumed as json at the end anyway, we go though an extra step?